kvserver: add timeout for sending Raft snapshots

erikgrinaker opened the pull request:
This patch adds a timeout of 20 minutes when sending Raft snapshots.
While a snapshot is in transit, Raft log truncation is halted to allow
the recipient to catch up.  If the snapshot takes very long to transfer
for whatever reason this can cause the Raft log to grow very large. We
therefore set a conservative timeout to eventually allow Raft log
truncation while avoiding snapshot starvation -- even if another
snapshot is sent immediately, this still allows truncation up to the new
snapshot index.

Resolves #76295.

Release note (bug fix): There is now a 20 minute timeout when sending
Raft snapshots, to avoid stalled snapshot transfers preventing Raft log
truncation, which can cause the Raft log to grow very large.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76346)
<!-- Reviewable:end -->

tbg reviewed (commented):
Good idea. Bit confused about the overlapping cancellations and left a suggestion to boil it down to just one.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76346#-:-MvYIfAi-T4Ox31HmO_q:bfdgba)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)

___
*[pkg/kv/kvserver/replica_command.go, line 1696 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76346#-MvYI42lCWD5YHg5NtK6:-MvYI42lCWD5YHg5NtK7:b-fra5k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6edf294565fd87c6d988e932158a62b497ee207/pkg/kv/kvserver/replica_command.go#L1696)):*
> ```Go
> 		// switching between StateSnapshot and StateProbe. Even if we worked through
> 		// these, it would be susceptible to future similar issues.
> 		err := contextutil.RunWithTimeout(
> ```

This doesn't seem necessary, since `(*Replica).sendSnapshot` already does the same.
___
*[pkg/kv/kvserver/replica_command.go, line 2451 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76346#-MvYHaqg4-t6byZcLGhT:-MvYHaqg4-t6byZcLGhU:b-wnxjsy) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6edf294565fd87c6d988e932158a62b497ee207/pkg/kv/kvserver/replica_command.go#L2451)):*
> ```Go
> 	}()
> 
> 	ctx, cancel := context.WithTimeout(ctx, sendSnapshotTimeout) // nolint:context
> ```

All methods that send snapshots go through here, so isn't this line the only one you really need?  Also, wouldn't it be better somewhat better (for error reporting) to wrap the call to `SendSnapshot` in this method in a `RunWithTimeout` instead?

The changes to `makeRateLimitedTimeoutFunc` also don't seem necessary since ultimately they call in here. It makes sense to point out (perhaps in the cluster setting and on the replicate queue itself next to setting up the timeout func) that the additional hard limit is in the mix, though.


<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76346)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @tbg)

___
*[pkg/kv/kvserver/replica_command.go, line 2451 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/76346#-MvYHaqg4-t6byZcLGhT:-MvYJM1S0AmQT2LuVrwC:btwpu4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c6edf294565fd87c6d988e932158a62b497ee207/pkg/kv/kvserver/replica_command.go#L2451)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

All methods that send snapshots go through here, so isn't this line the only one you really need?  Also, wouldn't it be better somewhat better (for error reporting) to wrap the call to `SendSnapshot` in this method in a `RunWithTimeout` instead?

The changes to `makeRateLimitedTimeoutFunc` also don't seem necessary since ultimately they call in here. It makes sense to point out (perhaps in the cluster setting and on the replicate queue itself next to setting up the timeout func) that the additional hard limit is in the mix, though.
</blockquote></details>

Oops, this shouldn't have been in here. I was going back and forth on whether I should just set a hard limit in the method itself or leave it to the caller to decide -- ended up leaving it to the caller, to separate policy from mechanism.  But it's probably fine to just do it here too, at least that way we won't have to worry about callers forgetting about it.


<!-- Sent from Reviewable.io -->

tbg commented:
I think I prefer dowing it just near the send snapshot boundary, because we already have a few callers and there might be more (think sending snaps from followers).

erikgrinaker committed:
kvserver: add timeout for sending Raft snapshots

This patch adds a timeout of 20 minutes when sending Raft snapshots.
While a snapshot is in transit, Raft log truncation is halted to allow
the recipient to catch up. If the snapshot takes very long to transfer
for whatever reason this can cause the Raft log to grow very large. We
therefore set a conservative timeout to eventually allow Raft log
truncation while avoiding snapshot starvation -- even if another
snapshot is sent immediately, this still allows truncation up to the new
snapshot index.

Release note (bug fix): There is now a 20 minute timeout when sending
Raft snapshots, to avoid stalled snapshot transfers preventing Raft log
truncation, which can cause the Raft log to grow very large.

erikgrinaker reviewed (commented):
> I think I prefer dowing it just near the send snapshot boundary, because we already have a few callers and there might be more (think sending snaps from followers).

Ok, moved it all the way down to the `SendSnapshot()` call. PTAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76346#-:-Mv_Pz0p1k_WCH13N_87:b129hrs)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Apologies for suggesting this only now, but another alternative to consider would be to cap the deadline here at 20 minutes:

https://github.com/cockroachdb/cockroach/blob/9a7b735b1282bbb3fb7472cc26a47d516a446958/pkg/kv/kvserver/replica_raft.go#L1486-L1489

by adding a `deadline: timeutil.Now().Add(20*time.Minute),` (and checking that this works as intended, etc).

I think what we have here is fine though.

> Reviewed 7 of 7 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76346#-:-MvslnvOADeerMhp-jhZ:bzc7x1)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @erikgrinaker)



<!-- Sent from Reviewable.io -->

erikgrinaker reviewed (commented):
That would allow log truncation to go ahead, but it wouldn't cancel the snapshot. I'll go with the context fix, it's simple and gets the job done.

TFTR!

bors r=tbg

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76346#-:-Mvu6TKuBPpoCOtlAbre:be8dck9)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @tbg)



<!-- Sent from Reviewable.io -->
